singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,420 @@
+        if (event != null) {
+          try {
+            value = serializer.serialize(event);
+            record =
+                new ProducerRecord<>(this.topic, getAlternatePartition(partitionInfoList.size()),
+                    null, value);
+            kafkaProducer.send(record, new KafkaProducerCallback(event));
+          } catch (TException e) {
+            LOG.debug("[{}] failed to construct ProducerRecord because of serialization exception.",
+                Thread.currentThread().getName(), e);
+            OpenTsdbMetricConverter
+                .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_SERIALIZATION_EXCEPTION, 1,
+                    "host=" + host, "stage=" + stage.toString(),
+                    "logName=" + event.getLoggingAuditHeaders().getLogName());
+            eventTriedCount.remove(event.getLoggingAuditHeaders());
+          }
+        }
+      } catch (InterruptedException e) {
+        LOG.warn("[{}] got interrupted when polling the queue and while loop is ended!",
+            Thread.currentThread().getName(), e);
+        OpenTsdbMetricConverter
+            .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_DEQUEUE_INTERRUPTED_EXCEPTION, 1,
+                "host=" + host, "stage=" + stage.toString());
+        break;
+      } catch (Exception e) {
+        LOG.warn("Exit the while loop and finish the thread execution due to exception: ", e);
+        OpenTsdbMetricConverter.incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_EXCEPTION, 1,
+            "host=" + host, "stage=" + stage.toString());
+        break;
+      }
+    }
+  }
+
+
+  public class KafkaProducerCallback implements Callback {
+
+    private LoggingAuditEvent event;
+
+    public KafkaProducerCallback(LoggingAuditEvent event) {
+      this.event = event;
+    }
+
+    public void checkAndEnqueueWhenSendFailed(RecordMetadata recordMetadata, Exception e) {
+      // if exception thrown (i.e. the send failed), the partition is added to badPartitions.
+      badPartitions.add(recordMetadata.partition());
+      OpenTsdbMetricConverter
+          .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_PARTITION_ERROR, 1,
+              "host=" + host, "stage=" + stage.toString(), "topic=" + topic,
+              "partition=" + recordMetadata.partition());
+
+      // retry the failed event by inserting it at the beginning of the deque.
+      // If number of tries reaches 3, meaning that 3 partitions have been tried sending to but
+      // still failed, this event is dropped.
+
+      Integer count = eventTriedCount.get(event.getLoggingAuditHeaders());
+      if (count == null){
+        eventTriedCount.put(event.getLoggingAuditHeaders(), 1);
+        OpenTsdbMetricConverter
+            .gauge(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_EVENTS_RETRIED,
+                eventTriedCount.size(), "host=" + host, "stage=" + stage.toString(),
+                "topic=" + topic);
+      } else {
```

Comment:
compline the `else { if ..` as  one `else if (..` ?

Reply:
made the change in the update of the PR